### PR TITLE
Work branch

### PR DESCRIPTION
**Pull request by chasegriswold**
Lab 10 Power branch. The README contains the recorded power draw values for the 5 different scenarios. The CICD is passing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,8 +79,8 @@ Measure the power consumption of the following scenarios. For each scenario, cre
...
 To eliminate any bias in the measurements, use an external LED with a known current value limiting resistor on a GPIO pin. Don't use the built in LED, which requires the wireless module to be enabled. If you want a serial connection, you need to use the UART pins rather than USB and access it with a serial converter.
 
-Blink an LED, using sleep_ms to delay iterations.
-Blink an LED, using FreeRTOS with a thread.
-Run a busy loop to get 100% CPU utilization
-Place #pragma GCC optimize ("O0") at the top of the file to disable compiler optimizations.
-Run a tight loop like the one below. Multiplication typically requires large amounts of silicon, and adders with a full carry chain causes lots of activity. I don't know if these will be the worst instructions for energy usage, but this should be a decent place to start.
+1) Blink an LED, using sleep_ms to delay iterations.
+2) Blink an LED, using FreeRTOS with a thread.
+3) Run a busy loop to get 100% CPU utilization
```
Comment (warnerjon12) [inline README.md:84]: What you have as 3), 4), and 5) are all one activity. 4) and 5) are notes suggesting how to do this. There are also 3 more scenarios (really 4 or 5 if you count using different frequencies):

1. Set up an interrupt to toggle a GPIO in response to an external signal (like in lab 5).
    1. In the main loop to invoke the wait for interrupt instruction.
    1. Drive the interrupt pin with the signal generator.
    1. Try a couple of frequencies between 1KHz and 1 MHz.
1. Run the sleep demo. https://github.com/raspberrypi/pico-playground/tree/master/sleep/hello_sleep/hello_sleep_alarm.c
    1. More details are in section 2.11.5.1 of the RP2040 datasheet.
    1. Read through the demo source and the library code to make sure that the code matches the setup description in the datasheet
    1. https://github.com/raspberrypi/pico-extras/blob/master/src/rp2_common/pico_sleep/sleep.c
1. Run the dormant demo. https://github.com/raspberrypi/pico-playground/tree/master/sleep/hello_dormant/hello_dormant_gpio.c
    1. More details are in section 2.11.5.2 of the RP2040 datasheet.

File: src/hello_freertos.c [hunk 2]
```diff
@@ -1,54 +1,213 @@
...
+#define BLINK_TASK_PRIORITY     ( tskIDLE_PRIORITY + 2UL ) // Blink task priority
+#define BLINK_TASK_PRIORITY2    ( tskIDLE_PRIORITY + 5UL ) // Blink task priority
+#define BLINK_TASK_STACK_SIZE configMINIMAL_STACK_SIZE     // Stack size for blink task
 
+// Global variables
+int count = 0;      // Counter to toggle LED
+bool on = false;    // Current LED state (on or off)
+
+// 1. Blink an LED using sleep_ms to delay iterations
 void blink_task(__unused void *params) {
```
Comment (warnerjon12) [inline src/hello_freertos.c:23]: If we really want to compare sleep_ms vs. vTaskDelay, we should run this case without the overhead of FreeRTOS

File: src/hello_freertos.c [hunk 3]
```diff
@@ -1,54 +1,213 @@
...
+int main(void) {
+    // Initialize standard I/O for printing messages
     stdio_init_all();
-    const char *rtos_name;
-    rtos_name = "FreeRTOS";
-    TaskHandle_t task;
-    xTaskCreate(main_task, "MainThread",
-                MAIN_TASK_STACK_SIZE, NULL, MAIN_TASK_PRIORITY, &task);
+
+    // Uncomment the desired task to run ---------
```
Comment (warnerjon12) [inline src/hello_freertos.c:117]: Using conditional compilation with macros and using cmake's `target_compile_definitions` to create separate executables from the same source file is probably a better alternative to doing this.

File: src/hello_freertos.c [hunk 4]
```diff
@@ -1,54 +1,213 @@
...
+                BLINK_TASK_STACK_SIZE, NULL, BLINK_TASK_PRIORITY, NULL);
+
+    // Start the FreeRTOS scheduler
     vTaskStartScheduler();
-    return 0;
+
+    return 0; // Should never reach here unless scheduler fails
 }
+
+// ----- Original Code to debug that was the first pushed to work_branch
```
Comment (warnerjon12) [inline src/hello_freertos.c:147]: I would move this to a debug branch to keep the work branch cleaner. In that case there is also no need to have both sections of code in either branch.

File: README.md
Comment (warnerjon12) [inline README.md:1]: Please include a brief description of each scenario here so it reads naturally for comparison without having to go somewhere else to interpret which is which. Please also add the power usage (I'd just multiply the current by 5 V and add it in parenthesis to what you have).

**Review by warnerjon12 — CHANGES_REQUESTED** [review]
This is a good start. See my specific comments for suggested changes.

**Comment by chasegriswold** [comment]
Hey Ashton, if you see this, this was the lab where my lab partner was out sick, and I followed older instructions before you updated the README with new instructions. This is the complete lab based on what we were reasonably able to accomplish, based on the original instructions.

**Comment by chasegriswold** [comment]
Jon, your comments are appreciated. I spoke with Ashton about extenuating circumstances, and the lab is complete based on my reasonable ability. I had older instruction set.

**Pull request merged**